igcollect/ping.py

```diff
@@ -4,83 +4,90 @@
 Copyright (c) 2020 InnoGames GmbH
 """
 
-import _thread
+# this script is used to collect ping data to evaluate availability and latency
+# of our network (this should replace smokeping in the longrun)
+#
+# Please Not that the package fping has to be installed on your Linux server!
+
+
 import argparse
 import subprocess
+import time
+
+from multiprocessing import Pool
+from os import cpu_count
 
-from time import time, sleep
 
 def parse_args():
     parser = argparse.ArgumentParser(
-        description='Ping colecctor script for Graphana ')
-    parser.add_argument('adress', metavar='URL', type=str, nargs='+',
-                        help='the address zu ping to')
+        description='Ping collector script for Graphite ')
+    parser.add_argument('hosts', metavar='h', nargs='+',
+                        help='the adresses to ping')
     parser.add_argument('-c', '--count', type=int, default=20,
-                        help='how many tims should the script tries to ping')
-    parser.add_argument('--prefix', default='ping',
-                        help='the path to the value in Graphana')
+                        help='how many times should the script tries to ping')
+    parser.add_argument('-p', '--prefix', default='ping',
+                        help='the path to the value in Graphite')
+    parser.add_argument('-t', '--timeout', type=int, default=300,
+                        help='time in milliseconds'
+                             ' till the timeout is retched')
     return parser.parse_args()
 
 
 def main():
     args = parse_args()
-    hostname = args.adress
-    count = args.count
-    prefix = args.prefix
-
-    i = 1
-    hosts = ''
-    for host in hostname:
-        hosts += host + ' '
-        if i % 10 == 0:
-            _thread.start_new_thread(ping, (prefix, hosts, count))
-            hosts = ''
-    if hosts != '':
-        ping(prefix, hosts, count)
-
-    while _thread._count() > 0:
-        sleep(1)
-
-
-def ping(prefix, hosts, count):
-    cmd = ('fping -B1 -q -C {} -p {} {}'.format(count, (20/count)*2000, hosts))
-    fping = subprocess.getoutput(cmd)
-    output = fping.split('\n')
+
+    with Pool(cpu_count() * 2) as p:
+        for i in range(len(args.hosts) % 10):
+            p.starmap(ping, [(args.prefix, args.hosts[i*10:i*10+9],
+                              args.count, args.timeout)])
+        p.starmap(ping, [(args.prefix,
+                          args.hosts[(len(args.hosts) % 10) * 10:
+                                     len(args.hosts) - 1],
+                          args.count, args.timeout)])
+
+
+def ping(prefix, hosts, count, timeout):
+    hosts_string = " ".join(hosts)
+    # -p calculates the time fping waits between single ping probes
+    cmd = ('fping -B1 -q -C {} -p {} -t {} {}'.format(count, (20/count) * 1800,
+                                                      timeout, hosts_string))
+    output = subprocess.getoutput(cmd).split('\n')
```

Comment:
Reviewing is quite hard without knowing the output. Please comment in a short example what the output could be, so it will get easier to follow up the next steps.

Reply:
servers.af-control_admin.system.ping.nvidia_com.ping10 162.41 1583400411
servers.af-control_admin.system.ping.nvidia_com.ping11 166.09 1583400411
servers.af-control_admin.system.ping.nvidia_com.ping12 165.95 1583400411
servers.af-control_admin.system.ping.nvidia_com.ping13 163.94 1583400411
servers.af-control_admin.system.ping.nvidia_com.ping14 164.13 1583400411
servers.af-control_admin.system.ping.nvidia_com.ping15 165.00 1583400411
servers.af-control_admin.system.ping.nvidia_com.ping16 200.77 1583400411
servers.af-control_admin.system.ping.nvidia_com.ping17 164.14 1583400411
servers.af-control_admin.system.ping.nvidia_com.ping18 164.89 1583400411
servers.af-control_admin.system.ping.nvidia_com.ping19 164.24 1583400411
servers.af-control_admin.system.ping.nvidia_com.ping20 162.62 1583400411
servers.af-control_admin.system.ping.nvidia_com.fails 0 1583400411
servers.af-control_admin.system.ping.nvidia_com.fails/1 0.0 1583400411
servers.af-control_admin.system.ping.nvidia_com.min 162.41 1583400411
servers.af-control_admin.system.ping.nvidia_com.max 200.77 1583400411
servers.af-control_admin.system.ping.nvidia_com.avg 166.914 1583400411

Reply:
amazon.com     : 105.94 102.03 104.78 106.39 102.64 101.89 103.49 102.30 101.32 100.77 102.95 99.68 101.57 102.56 102.27 101.69 101.83 102.35 112.64 102.49
google.de      : 16.90 20.21 18.16 15.86 17.35 17.51 19.87 21.28 18.99 17.88 22.26 18.72 17.36 17.82 18.45 17.38 17.27 20.58 20.71 24.57
bing.de        : 9.11 8.91 7.56 7.52 10.74 7.37 9.52 12.29 9.79 8.69 11.54 10.62 10.37 8.15 7.84 13.30 11.69 10.32 9.26 12.65
duckducogo.com : 18.45 19.54 18.33 18.27 20.35 18.62 20.38 24.27 21.24 20.27 19.30 22.88 23.32 19.42 19.32 20.39 28.69 25.49 21.20 20.33
microsoft.com  : - - - - - - - - - - - - - - - - - - - -
facebook.com   : 4.32 5.28 4.41 4.99 4.55 4.98 6.98 7.75 3.05 4.87 5.15 5.56 5.02 4.21 4.48 4.23 10.80 8.19 5.23 6.73
amazon.de      : 33.72 31.45 32.87 33.61 30.33 29.56 33.51 30.46 32.32 31.19 33.16 32.91 30.55 34.69 31.80 35.02 33.93 32.23 33.01 32.29
bing.com       : 11.50 8.37 10.34 8.53 10.55 10.00 9.73 8.78 8.83 8.15 9.86 9.55 8.57 9.34 8.56 11.63 11.19 11.22 11.99 12.07
facebook.de    : 5.71 5.76 5.23 2.57 6.33 3.99 5.59 6.26 8.04 5.93 5.76 5.82 6.36 11.90 4.84 7.35 5.93 8.32 8.34 7.14
nvidia.com     : 164.35 164.37 165.78 164.13 164.21 163.73 167.73 165.68 165.67 166.29 166.51 166.59 167.34 167.13 165.61 165.80 164.69 167.79 167.22 167.10

Reply:
Thx, how about to put one example line as comment into the code, so the next maintainer will directly know about it?
Like:
```
# example.com : 105.94 - 104.78
# Showing times for each ping. A ping timeout will be shown as '-'.
```